clash-lib/src/Clash/Rewrite/Util.hs

```diff
@@ -520,31 +521,48 @@ liftAndSubsituteBinders inScope toLift toKeep body = do
     else
       go subst2 inl
 
+-- | Determines whether a global binder is work free. Errors if binder does
+-- not exist.
+isWorkFreeBinder :: HasCallStack => Id -> RewriteMonad extra Bool
+isWorkFreeBinder bndr =
+  -- TODO: Will this get stuck for self-recursive binders?
+  makeCachedU bndr workFreeBinders $ do
+    bExprM <- lookupVarEnv bndr <$> Lens.use bindings
+    case bExprM of
+      Nothing -> error ("isWorkFreeBinder: couldn't find binder: " ++ showPpr bndr)
+      Just b -> isWorkFree (bindingTerm b)
+
 -- | Determine whether a term does any work, i.e. adds to the size of the circuit
 isWorkFree
   :: Term
-  -> Bool
+  -> RewriteMonad extra Bool
 isWorkFree (collectArgs -> (fun,args)) = case fun of
-  Var i            -> isLocalId i && not (isPolyFunTy (varType i))
-  Data {}          -> all isWorkFreeArg args
-  Literal {}       -> True
+  Var i ->
+    if | isPolyFunTy (varType i) -> pure False
+       | isLocalId i -> pure True
+       | otherwise -> isWorkFreeBinder i
```

Comment:
Wait... the arguments applied to a global function should be work-free as well. And actually, it's more complicated than that. For example, if the global functions is (isomorphic to) `const`, then we don't have to look at the work of the arguments; but for global functions (isomorphic to) `id` we do need to look at the arguments. I guess the best we can do is a heuristic, and just say the arguments have to be work-free. Given we already do something similar Lambda.

Reply:
Is that true for local ids too then?

Reply:
Yes, but for local function we just always assume they perform work, so we don't need to look at their arguments. That's why there's the `isPolyFunTy` guard.

Reply:
Alright, implemented the heuristic. Thanks.